File: Web/ProcessingTools.Web.Documents/Scripts/app/models/data/bio/taxonomy/taxon-rank.js
(function (window) {
    'use strict';
    var app, models;

    window.app = window.app || {};
    app = window.app;

    app.models = app.models || {};
    models = app.models;

    function TaxonRank(taxonName, rank) {

        taxonName = taxonName ? taxonName.replace(/\s+/g, '') : '';
        if (taxonName.length < 1) {
            throw 'Null or whitespace taxon name';
        }

        rank = rank ? rank.replace(/\s+/g, '').toLowerCase() : '';
        if (rank.length < 1) {
            throw 'Null or whitespace';
        }

        this.id = null;
        this.taxonName = taxonName;
        this.rank = rank;
    }

    TaxonRank.prototype.compare = function (taxon) {
        var self = this;
        return (self.taxonName === taxon.taxonName) && (self.rank === taxon.rank);
    };

    models.TaxonRank = TaxonRank;

}(window));

File: src/ProcessingTools/Data/ProcessingTools.Data.Models.Entity/Bio/Taxonomy/TaxonRank.cs
// <copyright file="TaxonRank.cs" company="ProcessingTools">
// Copyright (c) 2021 ProcessingTools. All rights reserved.
// </copyright>

namespace ProcessingTools.Data.Models.Entity.Bio.Taxonomy
{
    using System.Collections.Generic;
    using System.ComponentModel.DataAnnotations;
    using ProcessingTools.Common.Constants.Data.Bio.Taxonomy;

    /// <summary>
    /// Taxon rank entity.
    /// </summary>
    public class TaxonRank
    {
        private string name;

        /// <summary>
        /// Gets or sets the ID of the taxon rank entity.
        /// </summary>
        [Key]
        public int Id { get; set; }

        /// <summary>
        /// Gets or sets the name of the taxon rank.
        /// </summary>
        [Required(AllowEmptyStrings = false)]
        [MaxLength(ValidationConstants.MaximalLengthOfRankName)]
        public string Name { get => this.name; set => this.name = value?.ToLowerInvariant(); }

        /// <summary>
        /// Gets the collection of taxon names.
        /// </summary>
        public virtual ICollection<TaxonName> Taxa { get; private set; } = new HashSet<TaxonName>();
    }
}
